File: Examples/Ruby/exportConfigs.rb
#!/usr/bin/env ruby -W0

require 'jimson'
require 'uri'
require 'net/http'
require 'optparse'
require 'inifile'
require 'fileutils'

HOST='host'
USERNAME='username'
PASSWORD='password'
LAST_TIMESTAMP='lastConfigTimestamp'

def usageAndExit()
    puts 'Usage:'
    puts '  exportConfigs.py <output directory> <settings file>'
    exit
end

def main()

    output = ARGV[0]
    configFile = ARGV[1]
  
    if ARGV.empty?
        usageAndExit()
    end
  
    unless File.exist?(output)
        puts 'Output directory does not exist: ' + output
        exit
    end
    unless File.file?(configFile)
        puts 'Setting file does not exist: ' + configFile
        exit
    end

    config = IniFile.load(configFile)
  
    config.each_section { |section|
        host = config[section][HOST]
        user = config[section][USERNAME]
        password = config[section][PASSWORD]
    
        if config[section][LAST_TIMESTAMP]
            lastTimestamp = config[section][LAST_TIMESTAMP]
        else
            lastTimestamp = 0
        end
        
        begin
            lastTimestamp = exportConfigs(output, host, user, password, lastTimestamp)
            
            if lastTimestamp
              config[section][LAST_TIMESTAMP] = lastTimestamp
              config.write
            end 
        rescue => e
            puts e.to_s
        end
      }  
end


def createFilename(output, network, ipAddress, path, timestamp)
    localPath = path
    if localPath.index('/') == 0
        localPath = localPath[1..-1]
    end  

    return output + '/' + network + '/' + ipAddress + '/' + Time.at(timestamp/1000).strftime('%Y-%m-%d_%H-%M_') + localPath
end  

def millisToDate(millis)
    tmp = Time.at(millis / 1000).gmtime.strftime("%Y-%m-%d %H:%M:%S")
    return tmp 
end

def hash_to_query(hash)
  return URI.encode(hash.map{|k,v| "#{k}=#{v}"}.join("&"))
end 

def exportConfigs(output, host, user, password, lastTimestamp)
    uri = URI.parse("https://" + host)
    https = Net::HTTP.new(uri.host, uri.port)
    https.use_ssl = true
    https.verify_mode = OpenSSL::SSL::VERIFY_NONE
    response = https.start { |https|
      https.get("/rest?j_username=" + user + "&j_password=" + password)
    }
    cookie = response['set-cookie'].split(';', 2)[0] 
  
    netld = Jimson::Client.new("https://" + host + "/rest?j_username=" + user + "&j_password=" + password, 
              "Cookie" => cookie)
    
    configs = netld['Configuration.retrieveConfigsSince', lastTimestamp]
    configs.each { |config|
        lastTimestamp = config['lastChanged']
        timestamp = millisToDate(lastTimestamp)

        network = config['managedNetwork']
        ipAddress = config['ipAddress']
        path = config['path']
        
        filename = createFilename(output, network, ipAddress, path, lastTimestamp)

        directory = File.dirname filename
        unless File.exist?(directory)
            FileUtils.mkdir_p(directory)
        end
        
        params = {
                'op' => 'config',
                'ipAddress' => ipAddress,
                'managedNetwork' => network,
                'configPath' => path,
                'timestamp' => Time.at(config['lastChanged'] / 1000).gmtime.strftime("%Y-%m-%dT%H:%M:%S"),
                'j_username' => user,
                'j_password' => password
                }
        
        url = URI.parse("https://" + host + "/servlet/inventoryServlet?" + hash_to_query(params))
        response = https.start { |https|
          https.get(url,
            {"Cookie" => cookie, "Accept-Encoding" => "deflate;q=0.6"}
          )
        }
        
        file = open(filename, 'w')
        file.write(response.body)
        file.close()

        puts "Wrote " + filename
    }
        
    netld['Security.logoutCurrentUser']

    return lastTimestamp
end

if __FILE__ == $0
  main()
end

File: Examples/Perl/exportConfigs.pl
use LWP::UserAgent;
use HTTP::Cookies;
use JSON;
use JSON::RPC::Client2;
use URI::Escape;
use Data::Dumper;
use POSIX qw(strftime);
use Try::Tiny;
use Config::Tiny;
use DateTime;
use File::Basename 'basename', 'dirname';
use File::Path 'mkpath';

### Initial authentication
my $cookie_jar = HTTP::Cookies->new( ignore_discard => 1 );
my $ua = LWP::UserAgent->new(
    cookie_jar => $cookie_jar,
    ssl_opts   => { SSL_use_cert => 0, verify_hostname => 0 },
#    debug => 1 
);

my $HOST = 'host';
my $USERNAME = 'username';
my $PASSWORD = 'password';
my $LAST_TIMESTAMP = 'lastConfigTimestamp';

sub usage_and_exit
{
    die "Usage:\n  exportConfigs.pl <output directory> <settings file>\n";
}

sub main
{
    my @argv = @_;
    
    if ($#argv == -1)
    {
        usage_and_exit();
    }

	my $output = $argv[0];
	my $config_file = $argv[1];
	
    if ($output eq '')
    {
        usage_and_exit();
    }

    if($config_file eq '')
    {
        usage_and_exit();
    }

	my ($host, $user, $password, $last_session_end);

	my $config = Config::Tiny->read($config_file);

	foreach my $section (keys %{$config}) 
	{
		$host = $config->{$section}->{$HOST};
	    $user = $config->{$section}->{$USERNAME};
	    $password = $config->{$section}->{$PASSWORD};
	
	    if($config->{$section}->{$LAST_TIMESTAMP})
	    {
	    	$last_timestamp = $config->{$section}->{$LAST_TIMESTAMP};
	    }
	    else
	    {
	    	$last_timestamp = 0;
	    }
	    
	    try
	    {
	    	$last_timestamp = export_configs($output, $host, $user, $password, $last_timestamp);
	    	$config->{$section}->{$LAST_TIMESTAMP} = $last_timestamp if ($last_timestamp);
	    }
	    catch
	    {
	    	warn "caught error: $_";
	    }
	}
	
	$config->write($config_file);
}

sub millis_to_date
{
	my $millis = shift;
    my $tmp = strftime('%Y-%m-%d %H:%M:%S', gmtime($millis/1000));

    return $tmp;
}

sub create_filename
{
    my ($output, $network, $ip_address, $path, $timestamp) = @_;
    
    $local_path = $path;
    
    if (index $local_path, "/" == 0)
	{
		$local_path = substr($local_path, 1);
	}
    
    return $output . '/' . $network . '/' . $ip_address . '/' . strftime('%Y-%m-%d_%H-%M_', gmtime($timestamp/1000)) . $local_path;
}

sub export_configs
{
    my ($output, $host, $user, $password, $last_timestamp) = @_;
    my ($network, $ip_address, $path, $timestamp);
    
	my $netld = JSON::RPC::Client2->new("https://".$host."/rest?j_username=".uri_escape_utf8($user)."&j_password=".uri_escape_utf8($password), $ua);

    $configs = $netld->call('Configuration.retrieveConfigsSince', {'timestamp' =>$last_timestamp});

    foreach my $config (@$configs)
    {
        $last_timestamp = $config->{'lastChanged'};
        $timestamp = millis_to_date($last_timestamp);

        $network = $config->{'managedNetwork'};
        $ip_address = $config->{'ipAddress'};
        $path = $config->{'path'};
        
        $filename = create_filename($output, $network, $ip_address, $path, $last_timestamp);

        $directory = dirname $filename;
        unless (-d $directory)
        {
            mkpath($directory)
        }

        $params = "op=config"
                ."&ipAddress=".$ip_address
                ."&managedNetwork=".$network
                ."&configPath=".$path
                ."&timestamp=".strftime('%Y-%m-%dT%H:%M:%S', gmtime($config->{'lastChanged'}/1000))
                ."&j_username=".$user
                ."&j_password=".$password;

        my $url = 'https://'.$host.'/servlet/inventoryServlet?'.$params;
        my $request = HTTP::Request->new(GET => $url);
		my $response = $ua->request($request);
		
		if($response->is_success)
		{
			open my $fh, '>>', $filename or die "$!";
			binmode $fh;
			print $fh $response->decoded_content;
			close $fh;
		}
		else
		{
			die $response->status_line, "\n";
		}
        
        print "Wrote ".$filename."\n";
    }

	$netld->call("Security.logoutCurrentUser");

    return $last_timestamp;
}

main(@ARGV);
